encoding/int_encode.go

```diff
@@ -1,30 +1,188 @@
 package encoding
 
 import (
-	"errors"
 	"io"
+	"math"
+	"math/bits"
+
+	"github.com/xichen2020/eventdb/generated/proto/encodingpb"
+	"github.com/xichen2020/eventdb/x/proto"
+
+	bitstream "github.com/dgryski/go-bitstream"
+)
+
+var (
+	dictEncodingMaxCardinalityInt = 1 << 8
 )
 
 // IntEncoder encodes int values.
 type IntEncoder interface {
 	// Encode encodes a collection of ints and writes the encoded bytes to the writer.
 	// Callers should explicitly call `Reset` before subsequent call to `Encode`.
-	Encode(writer io.Writer, values ForwardIntIterator) error
+	Encode(writer io.Writer, valuesIt RewindableIntIterator) error
 
 	// Reset resets the encoder.
 	Reset()
 }
 
 // IntEnc is a int encoder.
-type IntEnc struct{}
+type IntEnc struct {
+	metaProto       encodingpb.IntMeta
+	dictionaryProto encodingpb.IntDictionary
+	buf             []byte
+}
 
 // NewIntEncoder creates a new int encoder.
-func NewIntEncoder(writer io.Writer) *IntEnc { return &IntEnc{} }
+func NewIntEncoder() *IntEnc {
+	return &IntEnc{
+		// Make buf at least big enough to hold Uint64 values.
+		buf: make([]byte, uint64SizeBytes),
+	}
+}
 
 // Encode encodes a collection of ints and writes the encoded bytes to the writer.
-func (enc *IntEnc) Encode(writer io.Writer, values ForwardIntIterator) error {
-	return errors.New("not implemented")
+func (enc *IntEnc) Encode(
+	writer io.Writer,
+	valuesIt RewindableIntIterator,
+) error {
+	// Determine whether we want to do table compression
+	var (
+		max  int
+		curr int
+		idx  int
+	)
+	min := math.MaxInt64
+	dict := make(map[int]int)
+	for valuesIt.Next() {
+		curr = valuesIt.Current()
+		// Only grow the dict map if we are below dictEncodingMaxCardinalityInt.
+		// This way we track if we've exceeded the max # of uniques.
+		if len(dict) <= dictEncodingMaxCardinalityInt {
+			if _, ok := dict[curr]; !ok {
+				dict[curr] = idx
+				idx++
+			}
+		}
+
+		// Track max/min.
+		max = int(math.Max(float64(curr), float64(max)))
+		min = int(math.Min(float64(curr), float64(min)))
+	}
+
+	if valuesIt.Err() != nil {
+		return valuesIt.Err()
+	}
+
+	// Rewind iteration.
+	valuesIt.Rewind()
+
+	switch {
+	case len(dict) <= dictEncodingMaxCardinalityInt:
+		// Table encode if there is less than 256 unique valuesIt.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DICTIONARY
+		// Min number of bits to encode each actual value into the table.
+		enc.metaProto.BytesPerDictionaryValue = int64(math.Ceil(float64(bits.Len(uint(max))) / 8.0))
+		// Min number of bits required to encode table indices.
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint(len(dict))))
+	default:
+		// Default to delta encoding.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DELTA
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint(max - min)))
```

Comment:
Probably should be `max - min + 1` since you need to encode all values in [`min`, `max`].

Reply:
I only need to encode the max possible delta which is `max - min`.

Reply:
Sure, but you'd need to encode all values between [0, max - min], and the total count of that is still `max - min + 1` values.

Reply:
Mmmm well if there are enough bits to represent the value of the largest delta (`max - min`) aren't there enough bits to represent every other value from that number to 0? Adding an extra bit is unnecessary here. 

Consider the simple case where:
```
max := 7
min := 4
numBits := bits.Len(uint(max - min))
log.Println("numBits:", numBits)
log.Printf("%b", 3)
log.Printf("%b", 2)
log.Printf("%b", 1)
log.Printf("%b", 0)

--- STDOUT ---
2018/12/09 20:01:22 numBits: 2
2018/12/09 20:01:22 11
2018/12/09 20:01:22 10
2018/12/09 20:01:22 1
2018/12/09 20:01:22 0
```

As long as there are enough bits to represent `max-min`, then there are enough bits to represent all other numbers below that as well.

Reply:
Yeah it seems `bits.Len(n)` returns the number of bits needed to represent numbers in [0,n] (i.e., n+1 numbers), so we are talking about the same thing here.

Reply:
Ahh kk, 👍.